### PR TITLE
Tests for Uri.php

### PR DESCRIPTION
**Pull request by daweeb**
Many functions have a Throw exception listed in phpdocu, but no test to trigger it.
Is this Wanted? @midahp
https://github.com/daweeb/Http/blob/06ff3334b051c556e2bd297cc7ee63efb20b0d0e/src/Uri.php#L363

https://github.com/daweeb/Http/blob/06ff3334b051c556e2bd297cc7ee63efb20b0d0e/src/Uri.php#L422

Done
-  Host
-  Port
- Scheme
- Fragment
- UserInfo
- remove comments
-  csfixer/phpformat
-  authority

### CODE REVIEW

File: test/UriTest.php [hunk 1]
```diff
@@ -52,4 +55,155 @@ public function testToString()
...
+
+    public function testWithValidPort()
+    {
+        $port = '56564';
+        $uri = new Uri();
+        $uri = $uri->withPort($port);
+        $this->assertEquals($uri->getPort(), $port);
+    }
+
+    public function testWithInvalidPort()
```
Comment (midahp) [inline test/UriTest.php]: Add a test for port <0. Then add a check in Uri.php.

File: test/UriTest.php [hunk 2]
```diff
@@ -52,4 +55,155 @@ public function testToString()
...
+        $this->expectException(InvalidArgumentException::class);
+        $uri = $uri->withPort($port);
+    }
+
+    public function testWithNullPort()
+    {
+        $port = '';
+        $uri = new Uri();
+        $uri = $uri->withPort($port);
+        $this->assertEquals($uri->getPort(), null);
```
Comment (midahp) [inline test/UriTest.php]: This is a good example where run into problems with `assertEquals`. This will succeed if `$uri->getPort()` returns 0 (which is what happens here). Use `assertSame` to check the type aswell.
In this case you should use `assertNull($uri->getPort())` though, because you already know you want a `null` to be returned.

File: test/UriTest.php [hunk 3]
```diff
@@ -52,4 +55,155 @@ public function testToString()
...
+        $uri = $uri->withHost($host);
+        $this->assertEquals($uri->getHost(), 'groupware');
+    }
+
+    public function testWithInvalidHost()
+    {
+        $host = 'group@ware';
+        $uri = new Uri();
+        $uri = $uri->withHost($host);
+        $this->assertEquals($uri->getHost(), $host);
```
Comment (midahp) [inline test/UriTest.php]: If you want to add this check, this test should expect an exception for this value. Then add that exception in Uri.php

File: test/UriTest.php [hunk 4]
```diff
@@ -52,4 +55,155 @@ public function testToString()
...
+    }
+
+    public function testWithNullHost()
+    {
+        $host = '';
+        $uri = new Uri();
+        $uri = $uri->withHost($host);
+        $this->assertEquals($uri->getHost(), null);
```
Comment (midahp) [inline test/UriTest.php]: You should use assertSame('', $uri->getHost());

File: test/UriTest.php [hunk 5]
```diff
@@ -52,4 +55,155 @@ public function testToString()
...
+    }
+
+    public function testWithPathRemovesFragmentString()
```
Comment (midahp) [inline test/UriTest.php]: This test already exists above.

File: test/UriTest.php [hunk 6]
```diff
@@ -52,4 +55,155 @@ public function testToString()
...
+        $uri = $uri->withFragment($fragment);
+        $this->assertEquals($uri->getFragment(), $fragment);
+    }
+
+    public function testWithEmpytFragmentString()
+    {
+        $fragment = '';
+        $uri = new Uri();
+        $uri = $uri->withFragment($fragment);
+        $this->assertEquals($uri->getFragment(), null);
```
Comment (midahp) [inline test/UriTest.php]: use `assertSame('', $uri->getFragment()`

File: test/UriTest.php [hunk 7]
```diff
@@ -52,4 +55,155 @@ public function testToString()
...
+        $uri = $uri->withScheme($scheme);
+        $this->assertEquals($uri->getScheme(), 'feed');
+    }
+
+    public function testWithSchemeEmpty()
+    {
+        $uri = new Uri();
+        $scheme = '';
+        $uri = $uri->withScheme($scheme);
+        $this->assertEquals($uri->getScheme(), null);
```
Comment (midahp) [inline test/UriTest.php]: use `assertSame('', $uri->getScheme())`

File: test/UriTest.php [hunk 8]
```diff
@@ -52,4 +55,155 @@ public function testToString()
...
+    public function testWithAuthorityHostEmpty()
+    {
+        $uri = new Uri();
+        $host = '';
+        $user = 'test';
+        $port = '12345';
+        $uri = $uri->withHost($host);
+        $uri = $uri->withUserInfo($user);
+        $uri = $uri->withPort($port);
+        $this->assertEquals($uri->getAuthority(), '');
```
Comment (midahp) [inline test/UriTest.php]: use `assertSame` here

**Review by midahp — CHANGES_REQUESTED** [review]
Please solve the mentioned problems and try to add the necessary checks in Uri.php.
Also try to add tests for the `with*` methods to make sure they don't modify the original object

File: test/UriTest.php (hunk 8 above)
Comment (daweeb) [reply test/UriTest.php]: done

File: test/UriTest.php (hunk 7 above)
Comment (daweeb) [reply test/UriTest.php]: done

File: test/UriTest.php (hunk 6 above)
Comment (daweeb) [reply test/UriTest.php]: done

File: test/UriTest.php (hunk 5 above)
Comment (daweeb) [reply test/UriTest.php]: removed

File: test/UriTest.php (hunk 4 above)
Comment (daweeb) [reply test/UriTest.php]: done

File: test/UriTest.php (hunk 1 above)
Comment (daweeb) [reply test/UriTest.php]: done

File: test/UriTest.php (hunk 2 above)
Comment (daweeb) [reply test/UriTest.php]: done, but hat to add a new if in withport function. There is probably a smarter way of doing this. I will add a link when pushed to repo. https://github.com/daweeb/Http/blob/942b837266d1043954177cb91d2c0ad1d75636c8/src/Uri.php#L383

File: test/UriTest.php (hunk 3 above)
Comment (daweeb) [reply test/UriTest.php]: I added a hostname check, please validate if correct use-case applies here ( FILTER_VALIDATE_URL  would be better here, i guess).  Again added a null check.

File: src/Uri.php [hunk 9]
```diff
@@ -339,6 +340,17 @@ public function withUserInfo($user, $password = null): self
      */
     public function withHost($host): self
     {
+        if ($host == null) {
```
Comment (midahp) [inline src/Uri.php]: Use `if (!$host)` instead

File: src/Uri.php [hunk 10]
```diff
@@ -339,6 +340,17 @@ public function withUserInfo($user, $password = null): self
...
+            $ret = clone $this;
+            $ret->host = '';
+
+            return $ret;
+        }
+
+        if (!filter_var($host, FILTER_VALIDATE_DOMAIN, FILTER_FLAG_HOSTNAME)) {
+            throw new InvalidArgumentException('Invalid Hostname, Valid characters for hostnames are ASCII(7) letters.');
```
Comment (midahp) [inline src/Uri.php]: This message is not correct right? Not all ASCII letters are valid.

File: src/Uri.php [hunk 11]
```diff
@@ -364,6 +376,17 @@ public function withHost($host): self
      */
     public function withPort($port): self
     {
+        if ($port >= 65536 || $port <0) {
+            throw new InvalidArgumentException('Invalid Portnumber, port must be between 1 and 65535');
```
Comment (midahp) [inline src/Uri.php]: Port 0 is valid.

File: src/Uri.php [hunk 12]
```diff
@@ -364,6 +376,17 @@ public function withHost($host): self
...
+        }
+
+        if ($port == null) {
```
Comment (midahp) [inline src/Uri.php]: use:
```
if (!is_null($port) {
    $port = (int) $port;
}
$ret = clone $this;
$ret->port = $port;
return $ret;
```

File: test/UriTest.php [hunk 13]
```diff
@@ -52,4 +55,154 @@ public function testToString()
         $uri = new Uri($url);
         $this->assertEquals($url, (string) $uri);
     }
+
+    public function testWithValidPort()
```
Comment (midahp) [inline test/UriTest.php:62]: It would probably be useful to test the edge cases with port = 0 and port = 65535

File: test/UriTest.php [hunk 14]
```diff
@@ -52,4 +55,154 @@ public function testToString()
...
+
+    public function testWithValidUpperHost()
+    {
+        $host = 'GroupWare';
+        $uri = new Uri();
+        $uri = $uri->withHost($host);
+        $this->assertEquals($uri->getHost(), 'groupware');
+    }
+
+    public function testWithInvalidHost()
```
Comment (midahp) [inline test/UriTest.php]: Add a test with invalid host that only has ascii characters. Like `test##Host`.
Add tests for valid hosts that use the `-` and the `.` characters.
Add a test for invalid host that starts with the `-` character.

File: test/UriTest.php [hunk 15]
```diff
@@ -52,4 +55,154 @@ public function testToString()
...
+    {
+        $host = 'group¹ware';
+        $uri = new Uri();
+        $this->expectException(InvalidArgumentException::class);
+        $uri = $uri->withHost($host);
+    }
+
+    public function testWithNullHost()
```
Comment (midahp) [inline test/UriTest.php]: I would call this `testWithEmptyHost` but its fine

**Review by midahp — CHANGES_REQUESTED** [review]
Added some comments.
Also try to add tests for the with* methods to make sure they don't modify the original object

File: test/UriTest.php (hunk 13 above)
Comment (daweeb) [reply test/UriTest.php:62]: good idea, done.

File: src/Uri.php (hunk 9 above)
Comment (daweeb) [reply src/Uri.php]: done

File: src/Uri.php (hunk 11 above)
Comment (daweeb) [reply src/Uri.php]: changed. sry.

File: src/Uri.php (hunk 12 above)
Comment (daweeb) [reply src/Uri.php]: done, with error
`1) Horde\Http\Test\UriTest::testWithNullPort Failed asserting that 0 is null`
is this wanted? In the PHPDoc it says 
`  A null value provided for the port is equivalent to removing the port information.`
so i thought it should be null not 0.

File: test/UriTest.php (hunk 15 above)
Comment (daweeb) [reply test/UriTest.php]: changed.

**Comment by ralflang** [comment]
You are confusing things.

0 is an integer and an invalid port number - you must error.
null is "no port number given" and valid.

Check for null first and skip the port range check if it is null.

File: test/UriTest.php (hunk 14 above)
Comment (daweeb) [reply test/UriTest.php]: Added.

File: src/Uri.php (hunk 10 above)
Comment (daweeb) [reply src/Uri.php]: changed, sry.

File: src/Uri.php (hunk 12 above)
Comment (daweeb) [reply src/Uri.php]: From Ralf:

0 is an integer and an invalid port number - you must error.
null is "no port number given" and valid.

Check for null first and skip the port range check if it is null.

changed it.

File: src/Uri.php [hunk 16]
```diff
@@ -364,6 +376,17 @@ public function withHost($host): self
...
     public function withPort($port): self
     {
+        if (is_null($port)) {
+            $ret = clone $this;
+            $ret->port = null;
+
+            return $ret;
+        }
+
+        if ($port >= 65536 || $port <= 0) {
```
Comment (midahp) [inline src/Uri.php:386]: I think you missunderstood. I meant the Exception message needs to  say `between 0 and 65535`. You actually made port 0 invalid in the if. Port 0 is a valid port (as far as I know).

File: test/UriTest.php [hunk 17]
```diff
@@ -52,4 +58,259 @@ public function testToString()
...
+
+    public function testWithValidPortMax()
+    {
+        $port = '65535';
+        $uri = new Uri();
+        $uri = $uri->withPort($port);
+        $this->assertEquals($uri->getPort(), $port);
+    }
+
+    public function testWithValidPortMin()
```
Comment (midahp) [inline test/UriTest.php:78]: This needs to be adjusted, because port 0 is a valid port. See other comment about `withPort`

File: src/Uri.php (hunk 12 above)
Comment (midahp) [reply src/Uri.php]: Yes I mentioned it in another comment. But my code suggestion should still work. You test just tested the wrong thing.

File: src/Uri.php (hunk 16 above)
Comment (midahp) [reply src/Uri.php:386]: My mistake.
